src/common/buffer.cc

```diff
@@ -399,183 +458,380 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
   };
 #endif
 
-#ifdef CEPH_HAVE_SPLICE
-  class buffer::raw_pipe : public buffer::raw {
-  public:
-    explicit raw_pipe(unsigned len) : raw(len), source_consumed(false) {
-      size_t max = get_max_pipe_size();
-      if (len > max) {
-	bdout << "raw_pipe: requested length " << len
-	      << " > max length " << max << bendl;
-	throw malformed_input("length larger than max pipe size");
-      }
-      pipefds[0] = -1;
-      pipefds[1] = -1;
 
-      int r;
-      if (::pipe(pipefds) == -1) {
-	r = -errno;
-	bdout << "raw_pipe: error creating pipe: " << cpp_strerror(r) << bendl;
-	throw error_code(r);
-      }
+#ifdef CEPH_HAVE_SPLICE
+/*
+ * Specialization of buffer::raw that implements zero-copy techniques.
+ * It is created for linux kernel. Uses syscalls and behaviours specific to it.
+ *
+ * Data is stored in series of linux pipes.
+ * If not specifically required by invoking \ref get_data(), data will remain in kernel space.
+ *
+ * \note For non-linux builds this class is not compiled and buffer::create_zero_copy produces regular buffer.
+ */
+class buffer::raw_pipe : public buffer::raw {
+private:
+  static constexpr int read_end = 0;
+  static constexpr int write_end = 1;
+  static constexpr int min_pipe_size = 1024*256; // < drop from using zero-copy if pipes are too small
+public:
+  struct pipe_elem
+  {
+    size_t size;
+    int fd;
+  };
+  std::vector<pipe_elem> pipe_read_elements;
+  int pipe_curr_write_fd; //this changes as writing progresses to next pipes
+  size_t total_content_size;
+  unsigned expected_len;
+
+  explicit raw_pipe(unsigned expected_len) : raw(0),
+      pipe_curr_write_fd(-1),
+      total_content_size(0),
+      expected_len(expected_len) {
+    update_max_pipe_size();
+    if (get_max_pipe_size()<min_pipe_size) {
+      data = new char[expected_len];
+      len = expected_len;
+      return;
+    }
+    int no_pipes = expected_len / get_max_pipe_size() + 1; //this is estimated #pipes to be needed
+    pipe_read_elements.reserve(no_pipes);
+    //we will have as much data as we read
+    len = 0;
+  }
 
-      r = set_nonblocking(pipefds);
-      if (r < 0) {
-	bdout << "raw_pipe: error setting nonblocking flag on temp pipe: "
-	      << cpp_strerror(r) << bendl;
-	throw error_code(r);
-      }
+  ~raw_pipe() {
+    if (data)
```

Comment:
I am aware that delete operator does check for nullptr.
However, I wanted here to stress that raw_pipe works in two modes:
- pipe mode
- buffer mode
This condition is an expression of "if you are in buffer mode, delete". 
Textual sugar for humans.